### PR TITLE
Np 2569 add pagination to query

### PR DESCRIPTION
**Pull request by torarnet**

### CODE REVIEW

File: src/main/java/no/unit/nva/cristin/projects/Constants.java [hunk 1]
```diff
@@ -27,4 +29,8 @@
     public static final String PROJECTS_PATH = "/project/";
     private static final String DOMAIN_NAME_ENV = "DOMAIN_NAME";
     public static final String DOMAIN_NAME = ENVIRONMENT.readEnv(DOMAIN_NAME_ENV);
+    public static final String PAGE = "page";
+    public static final String PAGE_NUMBER_ONE = "1";
```
Comment (axthosarouris) [inline src/main/java/no/unit/nva/cristin/projects/Constants.java:33]: Is this the first page or the second page?

Comment (torarnet) [reply src/main/java/no/unit/nva/cristin/projects/Constants.java:33]: first page

File: src/main/java/no/unit/nva/cristin/projects/FetchCristinProjects.java [hunk 2]
```diff
@@ -89,4 +115,12 @@ private boolean isValidCharacter(char c) {
             || c == CHARACTER_PERIOD;
     }
 
+    private boolean isInteger(String str) {
+        try {
+            Integer.parseInt(str);
+            return true;
+        } catch (NumberFormatException e) {
```
Comment (axthosarouris) [inline src/main/java/no/unit/nva/cristin/projects/FetchCristinProjects.java:122]: why not Exception?

File: src/test/java/no/unit/nva/cristin/projects/FetchCristinProjectsTest.java [hunk 3]
```diff
@@ -52,6 +58,12 @@
     private static final String TITLE_ILLEGAL_CHARACTERS = "abc123- ,-?";
     private static final String INVALID_JSON = "This is not valid JSON!";
     private static final String EMPTY_LIST_STRING = "[]";
+    private static final String PAGE_NUMBER_TWO = "2";
```
Comment (axthosarouris) [inline src/test/java/no/unit/nva/cristin/projects/FetchCristinProjectsTest.java:61]: SECOND_PAGE?

File: src/test/java/no/unit/nva/cristin/projects/FetchCristinProjectsTest.java [hunk 4]
```diff
@@ -213,8 +225,11 @@ void handlerReturnsProjectsWrapperWithAllMetadataButEmptyHitsArrayWhenNoMatchesA
 
     @Test
     void getsCorrectUriWhenCallingQueryProjectsUriBuilder() throws Exception {
-        assertEquals(new URI(QUERY_CRISTIN_PROJECTS_EXAMPLE_URI), cristinApiClientStub
-            .generateQueryProjectsUrl(Map.of(TITLE, RANDOM_TITLE, LANGUAGE_KEY, LANGUAGE_NB)));
+        Map<String, String> params = Map.of(TITLE, RANDOM_TITLE, LANGUAGE_KEY, LANGUAGE_NB, PAGE, PAGE_NUMBER_ONE,
```
Comment (axthosarouris) [inline src/test/java/no/unit/nva/cristin/projects/FetchCristinProjectsTest.java:228]: Can you put a new line on every key-value pair? Also, either use KEY for all parameter keys or do not use it at all

File: src/test/java/no/unit/nva/cristin/projects/FetchCristinProjectsTest.java [hunk 5]
```diff
@@ -258,6 +273,51 @@ void handlerReturnsInternalErrorWhenUriCreationFails() throws Exception {
         assertTrue(gatewayResponse.getBody().contains(ERROR_MESSAGE_SERVER_ERROR));
     }
 
+    @Test
+    void handlerReturnsCristinProjectsWhenQueryContainsPageParameter() throws Exception {
+        InputStream input = requestWithQueryParameters(
+            Map.of(TITLE, RANDOM_TITLE, LANGUAGE_QUERY_PARAMETER, LANGUAGE_NB, PAGE, PAGE_NUMBER_TWO));
```
Comment (axthosarouris) [inline src/test/java/no/unit/nva/cristin/projects/FetchCristinProjectsTest.java:279]: Here you use the LANGUAGE_QUERY_PARAMETER  and before the LANGUAGE_KEY. any reason? Preferably, do not replicate the query parameters. use the constants defined in main

File: src/test/java/no/unit/nva/cristin/projects/FetchCristinProjectsTest.java [hunk 6]
```diff
@@ -258,6 +273,51 @@ void handlerReturnsInternalErrorWhenUriCreationFails() throws Exception {
...
+        GatewayResponse<ProjectsWrapper> gatewayResponse = GatewayResponse.fromOutputStream(output);
+
+        assertEquals(HttpURLConnection.HTTP_OK, gatewayResponse.getStatusCode());
+        assertTrue(gatewayResponse.getBody().contains(URI_WITH_PAGE_NUMBER_VALUE_OF_TWO));
+    }
+
+    @Test
+    void handlerThrowsBadRequestWhenQueryHasInvalidPageParameter() throws Exception {
+        InputStream input = requestWithQueryParameters(
+            Map.of(TITLE, RANDOM_TITLE, LANGUAGE_QUERY_PARAMETER, LANGUAGE_NB, PAGE, TITLE_ILLEGAL_CHARACTERS));
```
Comment (axthosarouris) [inline src/test/java/no/unit/nva/cristin/projects/FetchCristinProjectsTest.java:290]: There is some inconsistency with the parameters, they can be XXXXX, or XXXXX_KEY or XXX_QUERY_PARAMETER.

File: src/test/java/no/unit/nva/cristin/projects/FetchCristinProjectsTest.java [hunk 7]
```diff
@@ -258,6 +273,51 @@ void handlerReturnsInternalErrorWhenUriCreationFails() throws Exception {
...
+        handler.handleRequest(input, output, context);
+        GatewayResponse<ProjectsWrapper> gatewayResponse = GatewayResponse.fromOutputStream(output);
+
+        assertEquals(HttpURLConnection.HTTP_BAD_REQUEST, gatewayResponse.getStatusCode());
+        assertTrue(gatewayResponse.getBody().contains(ERROR_MESSAGE_PAGE_VALUE_INVALID));
```
Comment (axthosarouris) [inline src/test/java/no/unit/nva/cristin/projects/FetchCristinProjectsTest.java:295]: asserThat(gatewayRsponse.getBody,containsString(ERROR_....) will give you a better report when the tests fails than just false.

File: src/test/java/no/unit/nva/cristin/projects/FetchCristinProjectsTest.java [hunk 8]
```diff
@@ -258,6 +273,51 @@ void handlerReturnsInternalErrorWhenUriCreationFails() throws Exception {
...
+
+    @Test
+    void handlerReturnsCristinProjectsWhenQueryContainsResultsParameter() throws Exception {
+        InputStream input = requestWithQueryParameters(
+            Map.of(TITLE, RANDOM_TITLE, LANGUAGE_QUERY_PARAMETER, LANGUAGE_NB, NUMBER_OF_RESULTS, TEN_RESULTS));
+        handler.handleRequest(input, output, context);
+        GatewayResponse<ProjectsWrapper> gatewayResponse = GatewayResponse.fromOutputStream(output);
+
+        assertEquals(HttpURLConnection.HTTP_OK, gatewayResponse.getStatusCode());
+        assertTrue(gatewayResponse.getBody().contains(URI_WITH_TEN_NUMBER_OF_RESULTS));
```
Comment (axthosarouris) [inline src/test/java/no/unit/nva/cristin/projects/FetchCristinProjectsTest.java:306]: ditto

File: src/test/java/no/unit/nva/cristin/projects/FetchCristinProjectsTest.java [hunk 9]
```diff
@@ -258,6 +273,51 @@ void handlerReturnsInternalErrorWhenUriCreationFails() throws Exception {
...
+    @Test
+    void handlerThrowsBadRequestWhenQueryHasInvalidResultsParameter() throws Exception {
+        InputStream input = requestWithQueryParameters(
+            Map.of(TITLE, RANDOM_TITLE, LANGUAGE_QUERY_PARAMETER, LANGUAGE_NB,
+                NUMBER_OF_RESULTS, TITLE_ILLEGAL_CHARACTERS));
+        handler.handleRequest(input, output, context);
+        GatewayResponse<ProjectsWrapper> gatewayResponse = GatewayResponse.fromOutputStream(output);
+
+        assertEquals(HttpURLConnection.HTTP_BAD_REQUEST, gatewayResponse.getStatusCode());
+        assertTrue(gatewayResponse.getBody().contains(ERROR_MESSAGE_NUMBER_OF_RESULTS_VALUE_INVALID));
```
Comment (axthosarouris) [inline src/test/java/no/unit/nva/cristin/projects/FetchCristinProjectsTest.java:318]: ditto

**Review by axthosarouris — CHANGES_REQUESTED** [review]

**Comment by axthosarouris** [comment]
If you do not like hamcrest you can use assertJ

File: src/main/java/no/unit/nva/cristin/projects/ErrorMessages.java [hunk 10]
```diff
@@ -23,4 +23,8 @@
         "Parameter 'title' is missing or invalid. "
             + "May only contain alphanumeric characters, dash, comma, period and whitespace";
     public static final String ERROR_MESSAGE_LANGUAGE_INVALID = "Parameter 'language' has invalid value";
+    public static final String ERROR_MESSAGE_PAGE_VALUE_INVALID = "Parameter 'page' has invalid value";
+    public static final String ERROR_MESSAGE_BACKEND_FAILED_WITH_STATUSCODE =
+        "Backend responded with status: %s when client called uri: %s";
```
Comment (brinxmat) [inline src/main/java/no/unit/nva/cristin/projects/ErrorMessages.java:28]: What is backend in this context? "Remote service" or "us"?

**Review by brinxmat — CHANGES_REQUESTED** [review]

File: src/test/java/no/unit/nva/cristin/projects/FetchCristinProjectsTest.java [hunk 11]
```diff
@@ -255,12 +274,65 @@ void handlerReturnsInternalErrorWhenUriCreationFails() throws Exception {
...
+    }
+
+    @Test
+    void handlerThrowsBadRequestWhenQueryHasInvalidPageParameter() throws Exception {
+        InputStream input = requestWithQueryParameters(Map.of(
+            TITLE, RANDOM_TITLE,
+            LANGUAGE, LANGUAGE_NB,
+            PAGE, TITLE_ILLEGAL_CHARACTERS));
+        handler.handleRequest(input, output, context);
+        GatewayResponse<ProjectsWrapper> gatewayResponse = GatewayResponse.fromOutputStream(output);
```
Comment (axthosarouris) [inline src/test/java/no/unit/nva/cristin/projects/FetchCristinProjectsTest.java:300]: I think this is a GatewayResponse<Problem>

File: src/test/java/no/unit/nva/cristin/projects/FetchCristinProjectsTest.java [hunk 12]
```diff
@@ -255,12 +274,65 @@ void handlerReturnsInternalErrorWhenUriCreationFails() throws Exception {
...
+    }
+
+    @Test
+    void handlerThrowsBadRequestWhenQueryHasInvalidResultsParameter() throws Exception {
+        InputStream input = requestWithQueryParameters(Map.of(
+            TITLE, RANDOM_TITLE,
+            LANGUAGE, LANGUAGE_NB,
+            NUMBER_OF_RESULTS, TITLE_ILLEGAL_CHARACTERS));
+        handler.handleRequest(input, output, context);
+        GatewayResponse<ProjectsWrapper> gatewayResponse = GatewayResponse.fromOutputStream(output);
```
Comment (axthosarouris) [inline src/test/java/no/unit/nva/cristin/projects/FetchCristinProjectsTest.java:326]: ditto

**Review by axthosarouris — APPROVED** [review]

**Review by brinxmat — APPROVED** [review]

**Pull request merged**